File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        
        newList = ListNode()
        cur = newList
        cur1, cur2 = list1, list2
        ## while list 1 and 2 are not empty
        while cur1 and cur2:
            ## compare the the cur1 and cur2's value
            ## put the smaller one into the newList
            ## then traverse to the next node
            if cur1.val < cur2.val:
                cur.next = cur1
                cur1 = cur1.next
            
            else:
                cur.next = cur2
                cur2 = cur2.next

            cur = cur.next  
                    
        ## After while loop
        ## Handle the node left eithier from list1 or list2

        if cur1:
            cur.next = cur1
        if cur2:
            cur.next = cur2
        
        return newList.next

File: Data Structures & Algorithms/climbing-stairs/submission-2.java
class Solution {
    public int climbStairs(int n) {
        
    //     int one = 1;
    //     int two = 1;

    //     for (int i = 0; i < n -1; i++) {
    //         int temp = one;
    //         one = one + two;
    //         two = temp;
    //     }
    //     return one;
    // }







        // n = 1 -> 1
        //  1 

        // n = 2 -> 2
        //  1 + 1
        //  2

        // n = 3 -> 3
        // 111
        // 12
        // 2 1

        // n = 4 -> 5
        // 1111
        // 112
        // 121
        // 211
        // 22

        // n = 5 -> 8
        // 11111 
        // 1112 
        // 122 

        // n = 5 = f(4) + f(3)
        // f(n) = f(n-1) + f(n-2)

        if (n == 1) {return 1;};
        if (n == 2) {return 2;};

        return climbStairs(n - 1) + climbStairs(n - 2);

        
    }
}
